src/Eloquent/Model.php

```diff
@@ -231,6 +242,28 @@ public function setAttribute($key, $value)
         return parent::setAttribute($key, $value);
     }
 
+    /** @inheritdoc */
+    protected function asDecimal($value, $decimals)
+    {
+        try {
+            $value = (string) BigDecimal::of((string) $value)->toScale((int) $decimals, RoundingMode::HALF_UP);
+
+            return new Decimal128($value);
+        } catch (BrickMathException $e) {
+            throw new MathException('Unable to cast value to a decimal.', previous: $e);
+        }
+    }
+
+    /** @inheritdoc */
+    public function fromJson($value, $asObject = false)
+    {
+        if (! is_string($value)) {
+            $value = Json::encode($value ?? '');
```

Comment:
Why not directly returning the value, instead of encoding and decoding?

The object to array conversion can be done with `array_walk_recursive`.

Reply:
To prevent passing value to `json_encode` every time on creating/updating actions.
Before
```php
Casting::query()->create(['jsonValue' => json_encode(['key' => 'value'])]);
```
After
```php
Casting::query()->create(['jsonValue' => ['key' => 'value']]);
```

Reply:
It's a good thing you don't force storage in json strings. 

To understand my point, here's what happens when the data is not encoded in JSON:
```php
if (! is_string($value)) {
    return  Json::decode(Json::encode($value ?? ''), ! $asObject);
}
```

Why not return it directly:
```php
if (! is_string($value)) {
    return $value;
}
```

Now I see that a custom json encoder can be used.